README
Python Postfix Policy for SPF (python-policy-spf) 1.0
Python based policy daemon for Postfix SPF checking
Tumgreyspf source
 Copyright © 2004-2005, Sean Reifschneider, tummy.com, ltd.
 <jafo@tummy.com>
python-policyd-spf changes
 Copyright © 2007-2012 Scott Kitterman <scott@kitterman.com>
<https://launchpad.net/pypolicyd-spf>
Documentation inputs:
 Copyright © 2004-2005, Sean Reifschneider, tummy.com, ltd.
 <jafo@tummy.com>
 2003-2004 Meng Weng Wong <mengwong@pobox.com> from postfix-policyd-spf-perl
 Copyright © 2007-2012 Scott Kitterman <scott@kitterman.com>

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

=================

This is python-policyd-spf, an external policy checker for the postfix mail
server.  It will use pyspf to check SPF records to determine if email should 
be rejected by your server.

To install from the tar.gz (is this software has been packaged for your 
distribution, the packaged version is recommended and should be installed using
your normal distribution packaging tools):

1.  Extract the package from the tarball (tar -xvvzf ...)
2.  Enter the package directory (cd ...)
3.  python setup.py build
4.  As root python setup.py install

It requires Python and the pyspf (python-spf) library version 2.0 or higher.
If pyspf not available through your packaging system, it can be downloaded
from:

http://sourceforge.net/projects/pymilter/

To use the optional RFC 5451 Authentication-Results header, the authres module
is also needed.  It can be downloaded from pypi or from:

https://launchpad.net/authentication-results-python

Nothing is configured by default, so this will not interact with Postfix until
it has been set up.

See man 1 policyd-spf for information on setting up and using this policy
server.

See man 5 policyd-spf.conf for configuration file information.



CHANGES
# Legend:
# --- = A new release
#   + = Added a feature (in a backwards compatible way)
#   ! = Changed something significant, or removed a feature
#   * = Fixed a bug, or made a minor improvement

--- 3.0.3 (2023-02-19)
  * Fix reference to QueueID in policydspfsupp.py (LP: #2007123), thanks to
    Wolfgang Karall-Ahlborn for the bug and the fix

--- 3.0.2 (2022-12-16)
  * Document limits of Hide_Receiver option when using the milter interface
  * Fix parsing of multiple value comma separated datasets
  * Fix InternalHosts configuration file parsing and data type conversions
  * Add documentation of milter only options to policyd-spf.conf (5)

--- 3.0.1 (2022-12-08)
  * Documentation cleanup
  ! Moved data files, removing /local component of paths, since it led to
    inappropriate installation paths with some pip options

--- 3.0.0 (2022-11-30)
  ! Replaced setuptools with flit for building/installing (packagers: this
    require changes)
    + Modern PEP 517/PEP 621 based build interface (requires flit 3.8)
    + Dependencies now specified
    + Distributed as both sdist and wheel
  * Use capitalized qtype for spf.DNSLookup calls (Fixes HELO_Whitelist test
    failues)
  ! Changed recipient tracking to only use recipient host name (per RFC 7208)
    in the policy server if per user processing is not being used and as a
    result, changed Hide_Receiver default to No since it will no longer have
    user specific personal information in most cases (LP: #1822679)
  + Added new option QueueID to include Postfix queue ID for normal policy
    server logging when configured for post-queue operations (LP: #1963816)
  * Fixed type definitions for UMask and IntHosts (LP: #1939598)
  * Fix milter macro types so they can be found (LP: #1986753)
  + Ship default pyspf-milter.conf in tarball
  + Add pyspf.milter.8 man page for milter interface documentation
  * Correct pyspf-milter.conf location in init script
  * Fix failures with non-UTF-8 Mail From addresses in the milter (also
    requires pymilter 1.0.5 or later)

--- 2.9.3 (2022-02-07)
  * Add mention in policyd-spf.conf (5) in the Lookup_Time and Whitelist_
    Lookup_Time entries that errors will occur if the policy server timeout
    limits exceed Postfix's smtpd_policy_service_timeout.  This is only
    relevant to the policy server interface.
  * Add more information about setting policyd-spf_time_limit to 
    policyd-spf.1.  This is only relevant to the policy server interface.
  * Fixup test suite since earthlink.com and some others have SPF records now
    (LP #1930132).
  * Remove stray leftover import of own_socketfile so milter variant will
    start (LP: #1856391), (Debian #1005110).

--- 2.9.2 (2019-11-22)
  * Add mention in policyd-spf.conf (5) in the TestOnly entry that to get both
    TestOnly behavior and no header field appended, Header_Type = None also
    needs to be set (LP: #1849994)
  * Milter: Move drop_privileges before Milter.runmilter and delete
    own_socketfile so that the milter interface runs as the correct user
    without race conditions about changing ownership of the socket file when
    it hasn't been created yet (When the milter is started, it will create the
    socket based on uMask, so we don't need to manually change it)

--- 2.9.1 (2019-10-06)
  * Use /run instead of /var/run
  * Fix-up sysv init so it works
  * Catch pyspf tracebacks so at least we don't die (thanks to Adi Pircalabu
    for both the report and the suggested fix) (LP: #1842005)
  * Correct over-writing of SPF identity by SPF reason for HELO checks and the
    reverse for Mail From (LP: #1822685)

--- 2.9.0 (2019-02-01)
  + Initial alpha release with pyspf-milter added
  * Use HOSTNAME for Authserv_Id lookup function from dkimpy-milter for both
    pyspf-milter and policyd-spf
  * Fixed url in setup to point to the spf-engine project

--- 2.1.0 (2018-06-17)
  ! Rename to SPF_Engine due to plan to add a milter front end in addition to
    current policy service front end
  ! Switched to use of setuptools and entry points (because its the future)
  ! Refactored common policy code into __init__.py

--- 2.0.2 (2017-12-14)
  * Fix treatment of No_Mail configuration parameter so that specifying
    No_Mail = False (the default) does not cause incorrect results (Thanks to
    David Jones on spf-devel for reporting)
  * Conditionally import authres is Header_Type is AR and raise an error if it
    is missing (sorry pep-8) to avoid cases where users change the config
    and suddenly it doesn't work for an example, see:
    https://bugzilla.redhat.com/show_bug.cgi?id=1208876
  * Update and correct Mail_From_pass_restriction description in
    policyd-spf.conf(5 ()
  * Update HELO checking default option in policyd-spf.conf(5) (LP: #1724107)
  * Note that SPF_Not_Pass is not consistent with RFC 7208 in the HELO
    checking section of policyd-spf.conf(5) - already documented for Mail From

--- 2.0.1 (2016-12-08)
  * Man page formatting and spelling corrections
  * Corrected default debug level (LP: #1647089)
  * Amplified loging level '-1' description
  * Forward port version 1.3.2 fixes for detection of missing Authserv_Id that
    were inadvertently not brought back to trunk

--- 2.0.0 (2016-12-02)
  ! No longer python2 compatible, minimum python3 version is 3.3 for ipaddress
  ! Removed support for use of ipaddr
  ! Changed default for HELO checking from SPF_Not_Pass to Fail (same as
    MailFrom) even though I think Not Pass makes more sense in order to
    still the complaints (Fedora, you can drop your sed call in the spec file
    now). (LP: #1571144)
  ! Changed default for Authserv-ID to use local hostname to provide a
    reasonable default Authserv-ID. (LP: #1575608)
  ! Increased minimum pyspf (python-spf) version to 2.0.9 so that Void_Limit
    is always available and used.
  ! Added new Hide_Receiver option to prevent accidental disclosure of BCC
    receivers and enabled it by default to maximize privacy.  (LP: #1394294)

policydspfsupp.py
# -*- coding: utf-8 -*-
#
#  Tumgreyspf
#  Copyright © 2004-2005, Sean Reifschneider, tummy.com, ltd.
#
#  pypolicyd-spf changes
#  Copyright © 2007-12 Scott Kitterman <scott@kitterman.com>
'''
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
'''

import syslog
import os
import sys
import string
import re
import stat


#  default values
defaultConfigData = {
        'debugLevel' : 1,
        'HELO_reject' : 'SPF_Not_Pass',
        'Mail_From_reject' : 'Fail',
        'PermError_reject' : 'False',
        'TempError_Defer'  : 'False',
        'skip_addresses' : '127.0.0.0/8,::ffff:127.0.0.0//104,::1//128',
        'defaultSeedOnly' : 1,
        'Header_Type' : 'SPF'
        }


#################################
class ConfigException(Exception):
    '''Exception raised when there's a configuration file error.'''
    pass


####################################################################
def _processConfigFile(filename = None, config = None, useSyslog = 1,
        useStderr = 0):
    '''Load the specified config file, exit and log errors if it fails,
    otherwise return a config dictionary.'''

    import policydspfsupp
    if config == None: config = policydspfsupp.defaultConfigData
    if filename != None:
        try:
            _readConfigFile(filename, config)
        except Exception, e:
            if useSyslog:
                syslog.syslog(e.args[0])
            if useStderr:
                sys.stderr.write('%s\n' % e.args[0])
            sys.exit(1)
    return(config)


#################
class ExceptHook:
   def __init__(self, useSyslog = 1, useStderr = 0):
      self.useSyslog = useSyslog
      self.useStderr = useStderr
   
   def __call__(self, etype, evalue, etb):
      import traceback
      tb = traceback.format_exception(*(etype, evalue, etb))
      tb = map(string.rstrip, tb)
      tb = string.join(tb, '\n')
      for line in string.split(tb, '\n'):
         if self.useSyslog:
            syslog.syslog(line)
         if self.useStderr:
            sys.stderr.write(line + '\n')


####################
def _setExceptHook():
    sys.excepthook = ExceptHook(useSyslog = 1, useStderr = 1)


###############################################################
commentRx = re.compile(r'^(.*)#.*$')
def _readConfigFile(path, configData = None, configGlobal = {}):
    '''Reads a configuration file from the specified path, merging it
    with the configuration data specified in configData.  Returns a
    dictionary of name/value pairs based on configData and the values
    read from path.'''

    debugLevel = configGlobal.get('debugLevel', 0)
    if debugLevel >= 5: syslog.syslog('readConfigFile: Loading "%s"' % path)
    if configData == None: configData = {}
    nameConversion = {
            'debugLevel' : int,
            'HELO_reject' : str,
            'Mail_From_reject' : str,
            'PermError_reject' : str,
            'TempError_Defer' : str,
            'Mail_From_pass_restriction' : str,
            'HELO_pass_restriction' : str,
            'Prospective' : str,
            'Whitelist' : str,
            'skip_addresses': str,
            'Domain_Whitelist' : str,
            'Domain_Whitelist_PTR': str,
            'No_Mail': str,
            'Reject_Not_Pass_Domains' : str,
            'Per_User' : str,
            'defaultSeedOnly' : int,
            'Header_Type' : str,
            'Authserv_Id' : str
            }

    #  check to see if it's a file
    try:
        mode = os.stat(path)[0]
    except OSError, e:
        syslog.syslog(syslog.LOG_ERR,'ERROR stating "%s": %s' % ( path, e.strerror ))
        return(configData)
    if not stat.S_ISREG(mode):
        syslog.syslog(syslog.LOG_ERR,'ERROR: is not a file: "%s", mode=%s' % ( path, oct(mode) ))
        return(configData)

    #  load file
    fp = open(path, 'r')
    while 1:
        line = fp.readline()
        if not line: break

        #  parse line
        line = string.strip(string.split(line, '#', 1)[0])
        if not line: continue
        data = map(string.strip, string.split(line, '=', 1))
        if len(data) != 2:
            if len(data) == 1:
                if debugLevel >= 1:
                    syslog.syslog('Configuration item "%s" not defined in file "%s"'
                        % ( line, path ))
            else:
                syslog.syslog('ERROR parsing line "%s" from file "%s"'
                    % ( line, path ))
            continue
        name, value = data

        #  check validity of name
        conversion = nameConversion.get(name)
        if conversion == None:
            syslog.syslog('ERROR: Unknown name "%s" in file "%s"' % ( name, path ))
            continue

        if debugLevel >= 5: syslog.syslog('readConfigFile: Found entry "%s=%s"'
                % ( name, value ))
        configData[name] = conversion(value)
    fp.close()
    
    return(configData)



policydspfuser.py
# -*- coding: utf-8 -*-
#
#  pypolicyd-spf
#  Copyright © 2010 Scott Kitterman <scott@kitterman.com>
'''
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 2 as published 
    by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.'''

import syslog
import re
import string
import os

###############################################################
commentRx = re.compile(r'^(.*)#.*$')
def readUserConfigFile(path, recipient, configData):
    '''Reads a configuration file from the specified path, merging it
    with the configuration data specified in configData for the identified
    recipient.  Returns updated configData for the user.'''

    debugLevel = configData.get('debugLevel', 0)
    if debugLevel >= 4: syslog.syslog('readConfigFile: Loading "%s"' % path)
    if configData == None: configData = {}
    nameConversion = {
            'debugLevel' : int,
            'HELO_reject' : str,
            'Mail_From_reject' : str,
            'PermError_reject' : str,
            'TempError_Defer' : str,
            'Mail_From_pass_restriction' : str,
            'HELO_pass_restriction' : str,
            'Prospective' : str,
            'Whitelist' : str,
            'skip_addresses': str,
            'Domain_Whitelist' : str,
            'Domain_Whitelist_PTR': str,
            'No_Mail': str,
            'Reject_Not_Pass_Domains' : str,
            'defaultSeedOnly' : int
            }

    #  check to see if it's a file
    try:
        mode = os.stat(path)[0]
    except OSError, e:
        syslog.syslog('ERROR stating "%s": %s' % ( path, e.strerror ))
        return(configData)

    #  load file
    fp = open(path, 'r')
    while 1:
        line = fp.readline()
        if not line: break

        #  parse line
        line = string.strip(string.split(line, '#', 1)[0])
        if not line: continue
        data = map(string.strip, string.split(line, ',', 1))
        user, value = data
        if user != recipient:
            peruser = False
            continue
        values = {}
        valuelist = string.split(value, '|')
        for valuepair in valuelist:
            key, item = string.split(valuepair, '=')
            values[key] = item
        for config in values.iteritems():
            #  check validity of name
            conversion = nameConversion.get(config[0])
            name, value = config
            if conversion == None:
                syslog.syslog('ERROR: Unknown name "%s" in file "%s"' % ( name, path ))
                continue
            if debugLevel >= 5: syslog.syslog('readConfigFile: Found entry "%s=%s"'
                % ( name, value ))
            configData[name] = conversion(value)
        peruser = True
    fp.close()

    return configData, peruser


def datacheck(configData, recipient):
    debugLevel = configData.get('debugLevel', 1)
    if debugLevel >= 3: syslog.syslog('Starting to process per-user settings')
    userdata = configData.get('Per_User')
    if debugLevel >= 3: syslog.syslog('User data: '+ str(userdata))
    usertype, userlocation = string.split(userdata, ',')
    if usertype == "text":
         if debugLevel >= 4: syslog.syslog('Reading per user data (type text) from:  "%s"' % userlocation)
         configData, peruser = readUserConfigFile(userlocation, recipient, configData)
         if debugLevel >= 4: syslog.syslog('Per-user settings for "%s": "%s"' % (recipient, str(configData)))

    return configData, peruser


spf_engine/milter_spf.py
#! /usr/bin/python3
# Derived from dkim-milter.py code:
# Author: Stuart D. Gathman <stuart@bmsi.com>
# Copyright 2007 Business Management Systems, Inc.
# This code is under GPL.  See COPYING for details.
# and:
# dkimpy-milter: A DKIM signing/verification Milter application
# Author: Scott Kitterman <scott@kitterman.com>
# Copyright 2018 Scott Kitterman
# spf-milter.py:
# Author: Scott Kitterman <scott@kitterman.com>
# Copyright 2019 Scott Kitterman
"""    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA."""

import sys
import syslog
import Milter
import authres
import os
import tempfile
import re
from Milter.utils import parse_addr, parseaddr
import spf_engine
import spf_engine.policydspfsupp as config
from spf_engine.util import drop_privileges
from spf_engine.policydspfsupp import _setExceptHook
from spf_engine.util import write_pid
from spf_engine.util import fold

__version__ = "3.0.0"
FWS = re.compile(r'\r?\n[ \t]+')


class spfMilter(Milter.Base):
    "Milter to check SPF.  Each connection gets its own instance."

    def __init__(self):
        self.mailfrom = None
        self.id = Milter.uniqueID()
        self.instance_dict = {'0':'init',}
        self.instance_dict.clear()
        self.data = {}

    @Milter.noreply
    def connect(self, hostname, unused, hostaddr):
        self.internal_connection = False
        self.external_connection = False
        self.hello_name = None
        # sometimes people put extra space in sendmail config, so we strip
        self.receiver = self.getsymval('j').strip()
        try:
            self.Authserv_Id = milterconfig['Authserv_Id']
        except:
            self.Authserv_Id = self.receiver
        if hostaddr and len(hostaddr) > 0:
            ipaddr = hostaddr[0]
            if milterconfig['IntHosts']:
                if milterconfig['IntHosts'].match(ipaddr):
                    self.internal_connection = True
        else:
            ipaddr = ''
        self.connectip = ipaddr
        self.data['client_address'] = ipaddr
        self.data['helo_name'] = hostname
        self.data['recipient'] = '<UNKNOWN>'
        if milterconfig.get('MacroList') and not self.internal_connection:
            macrolist = milterconfig.get('MacroList')
            for macro in macrolist:
                macroname = macro.split('|')[0]
                macroname = '{' + macroname + '}'
                macroresult = self.getsymval(macroname)
                if ((len(macro.split('|')) == 1 and macroresult) or macroresult
                        in macro.split('|')[1:]):
                    self.external_connection = True
                else:
                    self.internal_connection = True
        if self.internal_connection:
            connecttype = 'INTERNAL'
        else:
            connecttype = 'EXTERNAL'
        if milterconfig.get('debugLevel') >= 1:
            syslog.syslog("connect from {0} at {1} {2}"
                          .format(hostname, hostaddr, connecttype))
        return Milter.CONTINUE

    # multiple messages can be received on a single connection
    # envfrom (MAIL FROM in the SMTP protocol) seems to mark the start
    # of each message.
    def envfrom(self, f, *stri):
        # This gets rid of any non-UTF-8 in the localpart.
        # FIXME: Not great for l macro, not sure what we can do.
        f = str(bytes(f, encoding='utf-8', errors='replace'))[2:-1]
        if milterconfig.get('debugLevel') >= 2:
            syslog.syslog("mail from: {0} {1}".format(f, stri))
        self.mailfrom = f
        t = parse_addr(f)
        if len(t) == 2:
            t[1] = t[1].lower()
            domain = t[1]
        else:
            domain = 'localhost.localdomain'
        self.canon_from = '@'.join(t)
        self.arheaders = []
        self.arresults = []
        self.data['sender'] = self.canon_from 
        if (not self.internal_connection or self.external_connection) and self.connectip:
            return self.check_spf()
        return Milter.CONTINUE

    @Milter.noreply
    def header(self, name, val):
        lname = name.lower()
        if lname == 'authentication-results':
            self.arheaders.append(val)
        return Milter.CONTINUE

    @Milter.noreply
    def eoh(self):
        self.bodysize = 0
        return Milter.CONTINUE

    @Milter.noreply
    def body(self, chunk):        # copy body to temp file
        return Milter.CONTINUE

    def eom(self):
        if self.arresults:
            h = results=self.arresults[0]
            h = fold(str(h))
            if milterconfig.get('debugLevel') >= 2:
                syslog.syslog(str(h))
            name, val = str(h).split(': ', 1)
            self.addheader(name, val, 0)
        return Milter.CONTINUE


    def check_spf(self):
        peruser = False
        perusermilterconfig = []
        rejectAction = ''
        deferAction = ''
        if not self.data.get('recipient'):
            self.data['recipient'] = 'none'
        if milterconfig.get('debugLevel') >= 3: syslog.syslog('Config: %s' % str(milterconfig))
        #  run the checkers  {{{3
        checkerValue = None
        checkerReason = None
        checkerValue, checkerReason, self.instance_dict, iserror = \
                spf_engine._spfcheck(self.data, self.instance_dict, milterconfig,
                peruser, perusermilterconfig)

        TestOnly = milterconfig.get('TestOnly')
        if TestOnly == 0 and checkerValue != 'prepend':
            checkerValue = None
            checkerReason = None

        if milterconfig.get('SPF_Enhanced_Status_Codes') == 'No':
            rejectAction = ''
            deferAction = 'defer_if_permit'
        else:
            if iserror:
                rejectAction = '5.7.24'
                deferAction = '4.7.24'
            else:
                rejectAction = '5.7.23'
                deferAction = '4.7.24'

        #  handle results  {{{3
        if milterconfig.get('Header_Type') != 'None' and checkerValue != 'reject':
            if milterconfig.get('debugLevel') >= 3: syslog.syslog('{0} {1}'.format('Header text', checkerReason))
            self.arresults.append(checkerReason)
        if milterconfig.get('debugLevel') >= 3: syslog.syslog('Action: {0}: Text: {1} Reject action: {2}'.format(checkerValue, checkerReason, rejectAction))

        if checkerValue == 'reject':
            if milterconfig.get('debugLevel') >= 1: syslog.syslog('{0} {1}'.format(rejectAction, checkerReason))
            self.setreply(str(550), rejectAction, checkerReason)
            return Milter.REJECT

        elif checkerValue == 'prepend':
            if milterconfig.get('Prospective'):
                return Milter.CONTINUE
            else:
                if milterconfig.get('Header_Type') != 'None':
                    if milterconfig.get('debugLevel') >= 1: syslog.syslog('prepend {0}'.format(checkerReason))
                else:
                    if milterconfig.get('debugLevel') >= 1: syslog.syslog('Header field not prepended: {0}'.format(checkerReason))
                return Milter.CONTINUE

        elif checkerValue == 'defer':
            if milterconfig.get('debugLevel') >= 1: syslog.syslog('{0} {1}'.format(deferAction, checkerReason))
            self.setreply(str(450), deferAction, checkerReason)
            return Milter.TEMPFAIL

        elif checkerValue == 'result_only':
            return Milter.SKIP
        else:
            return Milter.CONTINUE


def main():
    # Ugh, but there's no easy way around this.
    global milterconfig
    configFile = '/usr/local/etc/python-policyd-spf/policyd-spf.conf'
    if len(sys.argv) > 1:
        if sys.argv[1] in ('-?', '--help', '-h'):
            print('usage: pyspf-milter [<configfilename>]')
            sys.exit(1)
        configFile = sys.argv[1]
    milterconfig = config._processConfigFile(filename=configFile)
    # Unlike policyd, milter interface only suppports authres
    if milterconfig.get('Header_Type') != 'None':
        milterconfig['Header_Type'] = 'AR'
    # Per user configurations not supported on milter interface
    milterconfig['Per_User'] = False
    facility = eval("syslog.LOG_{0}".format('MAIL'))
    syslog.openlog(os.path.basename(sys.argv[0]), syslog.LOG_PID, facility)
    _setExceptHook()
    pid = write_pid(milterconfig)
    Milter.factory = spfMilter
    Milter.set_flags(Milter.CHGHDRS + Milter.ADDHDRS)
    miltername = 'pyspf-filter'
    socketname = milterconfig.get('Socket')
    syslog.syslog('pyspf-milter started:{0} user:{1}'
                  .format(pid, milterconfig.get('UserID')))
    sys.stdout.flush()
    drop_privileges(milterconfig)
    Milter.runmilter(miltername, socketname, 240)

if __name__ == "__main__":
    main()
